21.md

```diff
@@ -0,0 +1,209 @@
+
+### Service-side first response
+
+`SERVICE` must alter its JSON response to the first callback to include a `currencies` field, as follows:
+
+```typescript
+type BaseResponse = {
+    tag: "payRequest",
+    metadata: string,
+    callback: string,
+    maxSendable: number,
+    minSendable: number
+};
+
+type Currency = {
+  code: string, // Code of the currency, used as an ID for it. E.g.: BRL
+  name: string, // Name of the currency. E.g.: Reais
+  symbol: string, // Symbol of the currency. E.g.: R$
+  decimals: number, // Number of decimal places. E.g.: 2
+  multiplier: number, // Number of millisatoshis per smallets unit of currency. E.g.: 5405.405
+  convertible?: bool // Whether the currency can be converted into.
+}
+
+type ExtendedResponse = BaseResponse & {
+  currencies: Currency[]
+}
+```
+
+```diff
+{
+  "tag": "payRequest",
+  "metadata": '[["text/plain","$kenu ⚡ bipa.app"]]',
+  "callback": "https://api.bipa.app/ln/request/invoice/kenu",
+  "maxSendable": 1000000000,
+  "minSendable": 1000,
++ "currencies": [
++   {
++     "code": "BRL",
++     "name": "Real",
++     "symbol": "R$",
++     "decimals": 2,
++     "multiplier": 5404.405,
++     "convertible": true
++   }
++ ]
+}
+```
+
+- The inclusion of the `currencies` field implies the support of this extension
+- The inclusion of a `currency` implies it can be used for denomination of an amount
+- The inclusion of a `convertible currency` implies the `SERVICE` can quote and guarantee a price for given currency
+- The price returned on this response is not guaranteed by the `SERVICE` and is subject to change
+- The `code` of a `currency` will be used as an identifier for the next request
+- The `code` and other information must be according to [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) for currencies that are included in it
+- The order of the `currencies` must be interpreted by the `WALLET` as the receiving user preference for a currency
+
+### Wallet-side second request
+
+Upon seeing the `currencies` field on the response from `SERVICE`, a `WALLET` may show the user it has the option of denominating the amount in different currencies or even of sending sats to be credited as the different currency for the receiver.
+
+After gathering the input from the user, it must be encoded into the callback as query parameters as follows.
+
+The most general case, when the user is denominating the amount in `CURRENCY1`, but wants it to be converted to `CURRENCY2` on `SERVICE`.
+`CURRENCY1` and `CURRENCY2` can be the same, which would the common use case for remittances.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY1>&convert=<CURRENCY2>
+
+In a more particular case, when user is denominating the amount in `CURRENCY`, but does not want to convert it or the service does not support conversion.
+```diff
+- <callback><?|&>amount=<milliSatoshi>
++ <callback><?|&>amount=<AMOUNT><CURRENCY>
```

Comment:
Maybe we should add a separator between `<AMOUNT><CURRENCY>`? This is an example of parsing implemented in Rust using the current spec:

```rust
fn parse_amount_and_maybe_currency(input: &str) -> anyhow::Result<(u64, Option<&str>)> {
    let index = input.find(char::is_alphabetic);

    Ok(if let Some(index) = index {
        let amount = input[..index].parse::<u64>()?;
        let currency = &input[index..];

        (amount, Some(currency))
    } else {
        let amount = input.parse::<u64>()?;

        (amount, None)
    })
}
```

Using a separator would make parsing a lot easier

Reply:
I would note that using a dot could lead to some weird bugs in practice if someone leaves decimals in the amount in a language like JS that doesn't distinguish between floats and ints. If we use something other than a period, it's a bit more explicit in the handling of bad values.

One more unfortunate JS thing here...

```
parseInt("100.BRL") == 100
parseInt("BRL.100") == NaN
```

I'm a bit worried that if we put the unit after the amount, nodejs lnurl servers will still parse it as a valid sats amount.